src/lib/runtime/mounts/userbinds/userbinds.c

```diff
@@ -93,22 +96,38 @@ int _singularity_runtime_mount_userbinds(void) {
                 }
             }
 
-
             singularity_message(DEBUG, "Checking if bind point is already mounted: %s\n", dest);
-            if ( check_mounted(dest) >= 0 ) {
-                singularity_message(WARNING, "Not mounting requested bind point (already mounted in container): %s\n", dest);
+            dupdest = strdup(dest);
+            if ( dupdest == NULL ) {
+                singularity_message(ERROR, "Failed to allocate memory\n");
+                ABORT(255);
+            }
+            if ( strcmp(dest, "/") == 0 ) {
+                singularity_message(WARNING, "Attempt to bind %s to container root filesystem, bind point ignored\n", source);
                 continue;
             }
+            for ( ptr = dupdest + 1; *ptr; ptr++ ) {
+                if ( *ptr == '/' ) {
+                    *ptr = '\0';
+                    if ( check_mounted(dupdest) >= 0 ) {
+                        singularity_message(WARNING, "Nested bind detected for %s, skip file/directory creation\n", dupdest);
+                        nested = 1;
+                        break;
+                    }
+                    *ptr = '/';
+                }
+            }
+            free(dupdest);
 
-            if ( ( is_file(source) == 0 ) && ( is_file(joinpath(container_dir, dest)) < 0 ) ) {
+            if ( nested == 0 && ( is_file(source) == 0 ) && ( is_file(joinpath(container_dir, dest)) < 0 ) ) {
                 if ( singularity_registry_get("OVERLAYFS_ENABLED") != NULL ) {
-                    char *dir = dirname(strdup(dest));
+                    dir = dirname(strdup(dest));
```

Comment:
There is no error check and you lose the pointer returned by strdup() which cannot be free'd.